### ISSUE TITLE
NewStruct in Exec-Engine gives insufficient feedback

### ISSUE THREAD

**Issue by stefjoosten** [AmpersandTarski/Ampersand#267]
I have an Ampersand-script, Statements.adl, consolidated in AmpersandTarski/ampersand-models@92baec6e4025cd620d5ce7c5ddf563eeda660338. It was compiled by Ampersand v3.1.0[master:ac300ab], build time: 05-Oct-15. This script calls the following Exec-Engine rule:

    RULE addFirstPlaceholder : template;head |- head;bind
    VIOLATION ( TXT "{EX} NewStruct;Prefix"
              , TXT ";head;Sequence;", SRC I[Sequence], TXT ";Prefix;_NEW"
              , TXT ";seq;Item;_NEW;Sequence;", SRC I[Sequence]
              , TXT ";bind;Prefix;_NEW;Variable;" , TGT I[Variable]
              )
    ROLE ExecEngine MAINTAINS addFirstPlaceholder

The following `CLASSIFY` statements cause the rule to be type-correct.

    CLASSIFY Prefix ISA Item
    CLASSIFY Variable ISA Item
After installing the database, there exists one Template with three variables in the database. It is called 'Matrijs1'. 
Then I want to make a Statement based on that Template.  So I press + and select Statement. Then I pick 'Matrijs1' as the template. This is where the Exec-Engine should kick in to react on violation of rule addFirstPlaceholder. 
The following screenshot was taken at precisely that moment.  It shows that the Exec-Engine rule has not been executed completely
![image](https://cloud.githubusercontent.com/assets/275768/12218036/313c200c-b713-11e5-99ff-5aaafa64c6d6.png)

@RieksJ , you know a lot about Exec-Engine rules. Do you have an idea about the cause of this phenomenon?

**Comment by RieksJ** [comment]
First of all, it would help if you provide the steps that I could follow in order to reproduce the bug. Given your issue, I would expect the bug to occur already when installing the initial population. That doesn't happen at my side. Since I cannot re-create the Ampersand-version that you use, I suggest you use the version of december 10th (that I use) and see if that resolves the problem.

Then, I think your analysis of that it has to do with CLASSIFY is correct, since the empty log line points to the statement:

    , TXT ";seq;Item;_NEW;Sequence;", SRC I[Sequence]

which should work given the statement `CLASSIFY Prefix ISA Item`.

**Comment by stefjoosten** [comment]
Rieks, you are right. Information to reproduce the error was missing.

After installing the database, there exists one Template with three variables in the database. It is called 'Matrijs1'. 
Then I want to make a Statement based on that Template.  So I press + and select Statement. Then I pick 'Matrijs1' as the template. This is where the Exec-Engine should kick in to react on violation of rule addFirstVariable. see screenshot

**Comment by RieksJ** [comment]
Did you try the other Ampersand.exe?

**Comment by stefjoosten** [comment]
oops: that was rule 'addFirstPlaceholder'.

**Comment by stefjoosten** [comment]
No,  not yet. I'm on the road, so not able right now.

**Comment by stefjoosten** [comment]
Could it be that 'delVarHead' compensates what 'addFirstPlaceholder' has iserted in seq? When that alternates, that would explain Error 500.

**Comment by stefjoosten** [comment]
I did the following experiment. I changed the code into:

    RULE addFirstPlaceholder : template;head |- head;bind
    VIOLATION ( TXT "{EX} NewStruct;Prefix"
              , TXT ";head;Sequence;", SRC I[Sequence], TXT ";Item;_NEW"
              , TXT ";seq;Item;_NEW;Sequence;", SRC I[Sequence]
              , TXT ";bind;Prefix;_NEW;Variable;" , TGT I[Variable]
              )
    ROLE ExecEngine MAINTAINS addFirstPlaceholder

Basically, I experimented with the order of the sub-InsPair statements inside the NewStruct. No matter how I order these, the empty line in the log file is consequently on the place of `seq`.

I also changed `Prefix` to `Item` in the line of `head`, because `head` is declared with type `[Sequence*Item]`. This too seems to make no difference.

**Comment by stefjoosten** [comment]
# Solution
After a hint from Michiel this morning, I discovered that the use of `NewStruct` is subject to the following rule:

If you say `{EX} NewStruct;Prefix`, every occurrence of `_NEW` should have type `Prefix`. This holds for every concept (here: `Prefix`). So the code now becomes:

    RULE addFirstPlaceholder : template;head |- head;bind
    VIOLATION ( TXT "{EX} NewStruct;Prefix"
              , TXT ";head;Sequence;", SRC I[Sequence], TXT ";Prefix;_NEW"
              , TXT ";seq;Prefix;_NEW;Sequence;", SRC I[Sequence]
              , TXT ";bind;Prefix;_NEW;Variable;" , TGT I[Variable]
              )
    ROLE ExecEngine MAINTAINS addFirstPlaceholder

I tried it and it works! Thank you @Michiel-s !

**Comment by stefjoosten** [comment]
@Michiel-s said he plans to look into the feedback given to the programmer. In this particular case, I got no helpful feedback from the Exec-Engine. Maybe future occurrences of violations of the NewStruct-rule can be prevented by better feedback. Michiel, would you close this issue when the feedback is in place?
